python/tvm/relay/op/_tensor_grad.py

```diff
@@ -130,3 +137,137 @@ def clip_grad(orig, grad):
+
+
+@register_gradient("sum")
+def sum_grad(orig, grad):
+    """Returns [broadcast_to_like(grad, x)]"""
+    return [broadcast_to_like(grad, orig.args[0])]
+
+
+@register_gradient("max")
+def max_grad(orig, grad):
+    """Returns the gradient of max"""
+    # Only support axis=0, since broadcasting orig to x behaves incorrectly
+    x, axis = orig.args[0], orig.attrs.axis
+    assert(axis is not None and len(axis) == 1 and int(axis[0]) == 0)
+    orig = broadcast_to_like(orig, x)
+    grad = broadcast_to_like(grad, x)
+    indicators = cast_like(equal(orig, x), grad)
+    count = reduce_sum(indicators, axis, True)
+    return [divide(indicators, count) * grad]
+
+
+@register_gradient("nn.softmax")
+def softmax_grad(orig, grad):
+    """Returns [(grad - sum(grad * orig, orig.attrs.axis, True)) * orig]"""
+    return [(grad - reduce_sum(grad * orig, orig.attrs.axis, True)) * orig]
+
+
+@register_gradient("negative")
+def negative_grad(orig, grad):
+    return [negative(broadcast_to_like(grad, orig.args[0]))]
+
+
+#x `dense` y = x . yT
+@register_gradient("nn.dense")
+def dense_grad(orig, grad):
+    # todo: check attrs
+    data, weight = orig.args
+    return [collapse_sum_like(op.nn.dense(grad, transpose(weight)), data),
+            collapse_sum_like(transpose(op.nn.dense(transpose(data), transpose(grad))), weight)]
+
+
+# UNTESTED
+@register_gradient("cast")
+def cast_grad(orig, grad):
+    grad = cast_like(grad, orig.args[0])
+    return [broadcast_to_like(grad, orig.args[0])]
+
+
+# UNTESTED
+@register_gradient("reshape")
+def reshape_grad(orig, grad):
+    return [reshape_like(grad, orig.args[0])]
+
+
+# UNTESTED
+@register_gradient("take")
+def take_grad(orig, grad):
+    x, y = orig.args
+    return [relay.zeros_like(x), relay.zeros_like(y)]
+    return [relay.Call(op.get("take_grad"), [x, y, grad], orig.attrs), zeros_like(y)]
+
+
+@register_gradient("shape_of")
+def shape_of_grad(orig, grad):
+    return [zeros_like(orig.args[0])]
+
+
+@register_gradient("nn.bias_add")
+def bias_add_grad(orig, grad):
+    """repeat the bias, then it is just add."""
+    # focuse on conv2d rn.
+    return [orig.args[0], orig.args[1]]
+
+
+@register_gradient("nn.batch_flatten")
+def batch_flatten_grad(orig, grad):
+    """reshape"""
+    # focuse on conv2d rn.
+    return [orig.args[0]]
+
+
+@register_gradient("nn.global_avg_pool2d")
+def global_avg_pool2d_grad(orig, grad):
+    """repeat the h w dimension"""
+    # focuse on conv2d rn.
+    return [orig.args[0]]
+
+
+@register_gradient("nn.batch_norm")
+def batch_norm_grad(orig, grad):
+    """multiply some stuff"""
+    # batchnorm has a wrong api so we will not waste time implementing it.
+    a, b, c, d, e = orig.args
+    return [a, b, c, d, e]
+
+
+@register_gradient("nn.conv2d")
+def conv2d_grad(orig, grad):
+    a, b = orig.args
+    x, y = relay.TupleWrapper(relay.Call(op.get("nn.conv2d_grad"), [a, b, grad], orig.attrs), 2)
+    return [a, b]
```

Comment:
Which of the grad implementations here are completed? A lot of these look like placeholders; e.g. the conv2d and pool2d grads aren't returning real gradients here.

Reply:
basically only entropy. this is mostly a WIP.

Reply:
It would be better to mark incomplete gradients as incomplete somehow, e.g. using a lower `plevel`.

Reply:
@sgrechanik-h it is ok, this is mostly for testing rn. I think I will open lots of small PR for the gradients.